File: agenkit-ts/examples/adapters/bedrock-basic.ts
/**
 * Basic Amazon Bedrock example
 *
 * Demonstrates:
 * - Bedrock adapter configuration
 * - Single-turn completion with Claude on Bedrock
 * - Multi-turn conversation
 * - Streaming responses
 * - AWS credential handling
 *
 * Setup:
 *   Configure AWS credentials (use IAM role, profile, or environment variables):
 *     export AWS_ACCESS_KEY_ID=your-key
 *     export AWS_SECRET_ACCESS_KEY=your-secret
 *     export AWS_REGION=us-east-1
 *   npm run build
 *   node dist/examples/bedrock-basic.js
 */

import { BedrockAdapter } from '../src/adapters/bedrock';
import { createMessage } from '../src/core/interfaces';

async function main() {
  console.log('='.repeat(60));
  console.log('AgentKit TypeScript - Amazon Bedrock Example');
  console.log('='.repeat(60));
  console.log();

  // Create Bedrock adapter
  const adapter = new BedrockAdapter({
    region: process.env.AWS_REGION || 'us-east-1',
    modelId: 'anthropic.claude-3-5-sonnet-20241022-v2:0',
    temperature: 0.7,
    maxTokens: 1024,
  });

  console.log(`✓ Initialized ${adapter.name}`);
  console.log(`✓ Capabilities: ${adapter.capabilities.join(', ')}`);
  console.log(`✓ Region: ${process.env.AWS_REGION || 'us-east-1'}`);
  console.log();

  // Example 1: Simple completion
  console.log('-'.repeat(60));
  console.log('Example 1: Simple Completion');
  console.log('-'.repeat(60));
  console.log();

  const message1 = createMessage({
    role: 'user',
    content: 'Explain the concept of cloud computing in one paragraph.',
  });

  console.log(`User: ${message1.content}`);
  console.log();

  const response1 = await adapter.process(message1);
  console.log(`Claude (Bedrock): ${response1.content}`);
  console.log();
  console.log('Metadata:');
  console.log(`  Model: ${response1.metadata?.model}`);
  console.log(`  Tokens: ${JSON.stringify(response1.metadata?.usage)}`);
  console.log(`  Stop reason: ${response1.metadata?.stop_reason}`);
  console.log();

  // Example 2: With system message
  console.log('-'.repeat(60));
  console.log('Example 2: Conversation with System Message');
  console.log('-'.repeat(60));
  console.log();

  const conversation = [
    createMessage({
      role: 'system',
      content: 'You are a helpful AWS expert who explains cloud services clearly.',
    }),
    createMessage({
      role: 'user',
      content: 'What is Amazon Bedrock?',
    }),
  ];

  console.log('System: You are a helpful AWS expert who explains cloud services clearly.');
  console.log('User: What is Amazon Bedrock?');
  console.log();

  const response2 = await adapter.complete(conversation);
  console.log(`Claude: ${response2.content}`);
  console.log();
  console.log(`Input tokens: ${response2.metadata?.usage?.prompt_tokens}`);
  console.log(`Output tokens: ${response2.metadata?.usage?.completion_tokens}`);
  console.log();

  // Example 3: Streaming response
  console.log('-'.repeat(60));
  console.log('Example 3: Streaming Response');
  console.log('-'.repeat(60));
  console.log();

  const message3 = createMessage({
    role: 'user',
    content: 'List the top 5 benefits of using Amazon Bedrock.',
  });

  console.log(`User: ${message3.content}`);
  console.log();
  console.log('Claude (streaming): ');

  let fullResponse = '';
  for await (const chunk of adapter.processStream(message3)) {
    process.stdout.write(chunk.content as string);
    fullResponse += chunk.content;
  }

  console.log();
  console.log();
  console.log(`Full response length: ${fullResponse.length} characters`);
  console.log();

  // Example 4: Different Bedrock models
  console.log('-'.repeat(60));
  console.log('Example 4: Using Claude 3 Haiku (faster, cheaper)');
  console.log('-'.repeat(60));
  console.log();

  const haikuAdapter = new BedrockAdapter({
    region: process.env.AWS_REGION || 'us-east-1',
    modelId: 'anthropic.claude-3-haiku-20240307-v1:0',
    temperature: 0.7,
    maxTokens: 200,
  });

  const message4 = createMessage({
    role: 'user',
    content: 'What is serverless computing in one sentence?',
  });

  console.log(`User: ${message4.content}`);
  console.log();

  const response4 = await haikuAdapter.process(message4);
  console.log(`Claude Haiku: ${response4.content}`);
  console.log();
  console.log('Available Bedrock models:');
  console.log('  anthropic.claude-3-5-sonnet-20241022-v2:0 - Most capable');
  console.log('  anthropic.claude-3-haiku-20240307-v1:0 - Fast and cost-effective');
  console.log('  meta.llama3-70b-instruct-v1:0 - Llama 3 70B');
  console.log('  mistral.mistral-large-2402-v1:0 - Mistral Large');
  console.log('  amazon.titan-text-premier-v1:0 - Amazon Titan');
  console.log();

  // Example 5: Multi-turn dialogue
  console.log('-'.repeat(60));
  console.log('Example 5: Multi-Turn Dialogue');
  console.log('-'.repeat(60));
  console.log();

  const dialogue = [
    createMessage({ role: 'user', content: 'What is Amazon S3?' }),
  ];

  console.log('User: What is Amazon S3?');

  const turn1 = await adapter.complete(dialogue);
  console.log(`Claude: ${turn1.content}`);
  console.log();

  dialogue.push(createMessage({ role: 'assistant', content: turn1.content as string }));
  dialogue.push(createMessage({ role: 'user', content: 'What are some common use cases?' }));

  console.log('User: What are some common use cases?');

  const turn2 = await adapter.complete(dialogue);
  console.log(`Claude: ${turn2.content}`);
  console.log();

  console.log(`Total tokens in conversation: ${turn2.metadata?.usage?.total_tokens}`);
  console.log();

  console.log('-'.repeat(60));
  console.log('✓ All examples completed successfully!');
  console.log('-'.repeat(60));
}

main().catch((error) => {
  console.error('Error:', error.message);
  console.error('');
  console.error('Make sure you have AWS credentials configured:');
  console.error('  - IAM role (for EC2/ECS/EKS)');
  console.error('  - AWS profile (~/.aws/credentials)');
  console.error('  - Environment variables (AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY)');
  console.error('');
  process.exit(1);
});

File: agenkit-cpp/examples/adapters/bedrock-basic.cpp
/**
 * @file bedrock-basic.cpp
 * @brief Basic Amazon Bedrock usage example
 *
 * Demonstrates:
 * - Bedrock adapter configuration
 * - Single-turn completion
 * - Multiple model comparison
 * - AWS credential configuration
 * - Error handling
 * - Metadata extraction
 *
 * Prerequisites:
 *   1. AWS SDK for C++ installed with bedrock-runtime component
 *   2. AWS credentials configured (one of):
 *      - Environment variables: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY
 *      - AWS credentials file: ~/.aws/credentials
 *      - IAM role (if running on EC2/ECS)
 *   3. AWS Bedrock model access enabled in your account
 *
 * Setup:
 *   # Option 1: Environment variables
 *   export AWS_ACCESS_KEY_ID=your-access-key
 *   export AWS_SECRET_ACCESS_KEY=your-secret-key
 *   export AWS_DEFAULT_REGION=us-east-1
 *
 *   # Option 2: AWS credentials file
 *   aws configure
 *
 *   # Build with Bedrock support
 *   cmake -B build -S . -DAGENKIT_HAS_AWS_SDK=ON
 *   cmake --build build
 *   ./build/examples/bedrock-basic
 *
 * NOTE: This example requires AWS SDK for C++ to be installed.
 * If not available, it will fail with a runtime error.
 */

#include "agenkit/adapters/bedrock_agent.hpp"
#include "agenkit/core/message.hpp"
#include <iostream>
#include <cstdlib>

using namespace agenkit;
using namespace agenkit::adapters;
using namespace agenkit::core;

void print_separator(const std::string& title = "") {
    std::cout << "\n";
    std::cout << std::string(60, '=') << "\n";
    if (!title.empty()) {
        std::cout << title << "\n";
        std::cout << std::string(60, '=') << "\n";
    }
    std::cout << "\n";
}

int main() {
    print_separator("AgentKit C++ - Amazon Bedrock Basic Example");

    std::cout << "Amazon Bedrock provides access to foundation models:\n";
    std::cout << "  • Anthropic Claude (3.5 Sonnet, 3 Opus, 3 Haiku)\n";
    std::cout << "  • Meta Llama (3.2 90B, 11B, 3B, 1B)\n";
    std::cout << "  • Mistral (Large, 7B Instruct)\n";
    std::cout << "  • Amazon Titan (Premier, Express, Lite)\n\n";

    std::cout << "NOTE: This requires AWS SDK for C++ to be installed.\n\n";

    try {
        // Example 1: Simple completion with Claude
        print_separator("Example 1: Simple Completion with Claude 3.5 Sonnet");

        BedrockConfig config;
        config.region = "us-east-1";
        config.model = BedrockModels::CLAUDE_3_5_SONNET_V2;
        config.temperature = 0.7;
        config.max_tokens = 1024;

        // Optional: Provide explicit credentials
        const char* access_key = std::getenv("AWS_ACCESS_KEY_ID");
        const char* secret_key = std::getenv("AWS_SECRET_ACCESS_KEY");
        if (access_key && secret_key) {
            config.access_key_id = access_key;
            config.secret_access_key = secret_key;
            std::cout << "✓ Using credentials from environment variables\n";
        } else {
            std::cout << "✓ Using AWS default credential chain\n";
        }

        BedrockAgent bedrock(config);

        std::cout << "Model: " << config.model << "\n";
        std::cout << "Region: " << config.region << "\n";
        std::cout << "User: Explain AWS Bedrock in two sentences.\n\n";

        auto msg = Message::with_text("user", "Explain AWS Bedrock in two sentences.");
        auto future = bedrock.process(std::move(msg));
        auto result = future.get();

        if (result.is_ok()) {
            const auto& response = result.unwrap();
            std::cout << "Claude: " << response.content_as_str() << "\n\n";

            // Show metadata
            std::cout << "Metadata:\n";
            const auto& meta = response.metadata();
            if (meta.contains("model")) {
                std::cout << "  Model: " << meta["model"].get<std::string>() << "\n";
            }
            if (meta.contains("finish_reason")) {
                std::cout << "  Finish reason: " << meta["finish_reason"].get<std::string>() << "\n";
            }
            if (meta.contains("usage")) {
                const auto& usage = meta["usage"];
                if (usage.contains("total_tokens")) {
                    std::cout << "  Total tokens: " << usage["total_tokens"].get<int>() << "\n";
                }
            }
        } else {
            std::cerr << "Error: " << result.unwrap_err().message() << "\n";
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << "\n";
        std::cerr << "\nPlease ensure:\n";
        std::cerr << "  1. AWS SDK for C++ is installed\n";
        std::cerr << "  2. AWS credentials are configured\n";
        std::cerr << "  3. Bedrock model access is enabled in your AWS account\n";
        return 1;
    }

    // Example 2: Model comparison
    print_separator("Example 2: Model Comparison");

    try {
        std::string prompt = "Write a haiku about cloud computing.";

        std::vector<std::pair<std::string, std::string>> models = {
            {"Claude 3.5 Sonnet", BedrockModels::CLAUDE_3_5_SONNET_V2},
            {"Claude 3 Haiku", BedrockModels::CLAUDE_3_HAIKU},
            {"Llama 3.2 11B", BedrockModels::LLAMA_3_2_11B}
        };

        for (const auto& [name, model_id] : models) {
            BedrockConfig config;
            config.region = "us-east-1";
            config.model = model_id;
            config.max_tokens = 100;

            BedrockAgent bedrock(config);

            std::cout << "Model: " << name << "\n";
            std::cout << "User: " << prompt << "\n\n";

            auto msg = Message::with_text("user", prompt);
            auto future = bedrock.process(std::move(msg));
            auto result = future.get();

            if (result.is_ok()) {
                std::cout << "Response:\n" << result.unwrap().content_as_str() << "\n\n";
            } else {
                std::cout << "Error: " << result.unwrap_err().message() << "\n\n";
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << "\n";
    }

    // Example 3: Temperature variation
    print_separator("Example 3: Temperature Effects");

    try {
        std::string prompt = "Generate a creative company name for a tech startup.";

        for (double temp : {0.2, 0.7, 1.0}) {
            BedrockConfig config;
            config.region = "us-east-1";
            config.model = BedrockModels::CLAUDE_3_HAIKU;
            config.temperature = temp;
            config.max_tokens = 50;

            BedrockAgent bedrock(config);

            std::cout << "Temperature: " << temp << "\n";
            std::cout << "User: " << prompt << "\n\n";

            auto msg = Message::with_text("user", prompt);
            auto future = bedrock.process(std::move(msg));
            auto result = future.get();

            if (result.is_ok()) {
                std::cout << "Response: " << result.unwrap().content_as_str() << "\n\n";
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << "\n";
    }

    // Example 4: Using inference configuration
    print_separator("Example 4: Advanced Configuration");

    try {
        BedrockConfig config;
        config.region = "us-east-1";
        config.model = BedrockModels::CLAUDE_3_5_SONNET_V2;
        config.temperature = 0.8;
        config.max_tokens = 200;
        config.top_p = 0.9;
        config.stop_sequences = {"\n\n", "END"};

        BedrockAgent bedrock(config);

        std::cout << "Configuration:\n";
        std::cout << "  Model: " << config.model << "\n";
        std::cout << "  Temperature: " << config.temperature.value() << "\n";
        std::cout << "  Max tokens: " << config.max_tokens.value() << "\n";
        std::cout << "  Top P: " << config.top_p.value() << "\n";
        std::cout << "  Stop sequences: ";
        for (const auto& seq : config.stop_sequences) {
            std::cout << "\"" << seq << "\" ";
        }
        std::cout << "\n\n";

        std::cout << "User: Tell me a short story about AI.\n\n";

        auto msg = Message::with_text("user", "Tell me a short story about AI.");
        auto future = bedrock.process(std::move(msg));
        auto result = future.get();

        if (result.is_ok()) {
            std::cout << "Response: " << result.unwrap().content_as_str() << "\n\n";
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << "\n";
    }

    // Example 5: Using different regions
    print_separator("Example 5: Multi-Region Support");

    try {
        std::vector<std::string> regions = {"us-east-1", "us-west-2", "eu-west-1"};

        for (const auto& region : regions) {
            BedrockConfig config;
            config.region = region;
            config.model = BedrockModels::CLAUDE_3_HAIKU;
            config.max_tokens = 50;

            std::cout << "Region: " << region << "\n";
            std::cout << "Testing connectivity...\n";

            // Note: Model availability varies by region
            std::cout << "(Model availability varies by region)\n\n";
        }
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << "\n";
    }

    // Example 6: Model constants
    print_separator("Example 6: Available Model Constants");

    std::cout << "Anthropic Claude models:\n";
    std::cout << "  • " << BedrockModels::CLAUDE_3_5_SONNET_V2 << "\n";
    std::cout << "  • " << BedrockModels::CLAUDE_3_5_SONNET << "\n";
    std::cout << "  • " << BedrockModels::CLAUDE_3_OPUS << "\n";
    std::cout << "  • " << BedrockModels::CLAUDE_3_HAIKU << "\n\n";

    std::cout << "Meta Llama models:\n";
    std::cout << "  • " << BedrockModels::LLAMA_3_2_90B << "\n";
    std::cout << "  • " << BedrockModels::LLAMA_3_2_11B << "\n";
    std::cout << "  • " << BedrockModels::LLAMA_3_2_3B << "\n\n";

    std::cout << "Mistral models:\n";
    std::cout << "  • " << BedrockModels::MISTRAL_LARGE_2407 << "\n";
    std::cout << "  • " << BedrockModels::MISTRAL_7B << "\n\n";

    std::cout << "Amazon Titan models:\n";
    std::cout << "  • " << BedrockModels::TITAN_TEXT_PREMIER << "\n";
    std::cout << "  • " << BedrockModels::TITAN_TEXT_EXPRESS << "\n";

    print_separator("✓ All examples completed!");

    std::cout << "Key Features Demonstrated:\n";
    std::cout << "  • Bedrock adapter configuration\n";
    std::cout << "  • AWS credential management\n";
    std::cout << "  • Single-turn completion\n";
    std::cout << "  • Model comparison (Claude, Llama, Mistral)\n";
    std::cout << "  • Temperature effects\n";
    std::cout << "  • Advanced configuration (top_p, stop sequences)\n";
    std::cout << "  • Multi-region support\n";
    std::cout << "  • Error handling and metadata\n\n";

    std::cout << "Next Steps:\n";
    std::cout << "  • Enable model access in AWS Bedrock console\n";
    std::cout << "  • Try different models by changing config.model\n";
    std::cout << "  • Experiment with temperature and max_tokens\n";
    std::cout << "  • Use stop sequences for structured output\n";
    std::cout << "  • Deploy to AWS for IAM role-based auth\n";
    std::cout << "  • See https://docs.aws.amazon.com/bedrock/ for more info\n\n";

    return 0;
}
